### DISCUSSION TITLE
Query Java classpath for build targets

### DISCUSSION THREAD

**Discussion by knutwannheden** [bazelbuild/bazel#18769] [Q&A] [answered]
Hi!

I am new to Bazel and I am looking into integrate a third party tool as a rule (or extension?), so that it can be used in Bazel workspaces which have been configured appropriately. This tool requires information about the complete runtime classpath required to compile a given Java target, including the JDK source and target compatibility. Is there an API or a query which can be executed to obtain this information, including the paths in the local file system to any external jar files and the directory containing the class files produced for dependent Java targets? Also, I would like to identify which source files represent production source files and which are test sources.

I am sorry if I am not using the proper Bazel terminology here. I am still trying to learn about the build system...

**Comment by meteorcloudy** [comment, marked as answer]
I think you can introduce the external tool with an external repository (Read https://bazel.build/docs/external for more), and write a custom [build rule](https://bazel.build/extending/rules) to create actions using the tool with necessary information.

**Reply by knutwannheden** [reply]
That sounds good, thanks for the links! I will read the pages you mentioned. Any more details on how exactly to get a hold of the compile-time Java classpath entries (in the file system), which I can feed the external tool with?

**Reply by meteorcloudy** [reply]
Try find if any API in https://bazel.build/rules/lib/toplevel/java_common provides information you need ;)

**Reply by knutwannheden** [reply]
Thanks for the pointer! I will try to follow a tutorial on creating an external repository to see if I can figure it out.
